pkg/controlplane/apiserver/config.go

```diff
@@ -159,13 +166,38 @@ func BuildGenericConfig(
 		genericConfig.FlowControl, lastErr = BuildPriorityAndFairness(s, clientgoExternalClient, versionedInformers)
 	}
 
+	if reconcilers.PeerProxyFeatureEnabled() {
+		genericConfig.PeerEndpointReconciler = createPeerEndpointLeaseReconciler(*genericConfig, storageFactory)
+		genericConfig.PeerProxy, lastErr = BuildPeerProxy(versionedInformers, genericConfig.StorageVersionManager, s.ProxyClientCertFile,
+			s.ProxyClientKeyFile, s.GenericServerRunOptions.PeerCAFile, s.GenericServerRunOptions.PeerBindAddress)
+	}
+
 	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AggregatedDiscoveryEndpoint) {
 		genericConfig.AggregatedDiscoveryGroupManager = aggregated.NewResourceManager("apis")
 	}
 
 	return
 }
 
+// createPeerEndpointLeaseReconciler creates a clone of EndpointReconciler to set up a separate apiserver endpoint
+// reconciliation loop that is used to find network locations of peer apiservers for peer proxy
+func createPeerEndpointLeaseReconciler(c genericapiserver.Config, storageFactory serverstorage.StorageFactory) reconcilers.PeerEndpointReconciler {
+	endpointClient := corev1client.NewForConfigOrDie(c.LoopbackClientConfig)
+	endpointSliceClient := discoveryclient.NewForConfigOrDie(c.LoopbackClientConfig)
+	endpointsAdapter := reconcilers.NewEndpointsAdapter(endpointClient, endpointSliceClient)
+
+	ttl := controlplane.DefaultEndpointReconcilerTTL
+	config, err := storageFactory.NewConfig(api.Resource("apiServerIPInfo"))
+	if err != nil {
+		klog.Fatalf("Error creating storage factory config: %v", err)
+	}
+	masterLeases, err := reconcilers.NewPeerEndpointLeases(config, "/peer-serverleases/", ttl)
+	if err != nil {
+		klog.Fatalf("Error creating leases: %v", err)
+	}
+	return reconcilers.NewPeerEndpointLeaseReconciler(endpointsAdapter, masterLeases)
```

Comment:
we only need a subsset if this reconciler 

https://github.com/kubernetes/kubernetes/blob/00b085ad4ad0c5d26466667a32a28046ffacd7f1/pkg/master/reconcilers/lease.go#L109-L131

we don't need the Endpoints/EndpointSlices logic, only the leases are enough since we want to pass the ips information through them, The annotation will make this information publicly available

Reply:
Updated